### PR TITLE
Support arbitrary `X` in `data`

### LINKED ISSUE #86: Make less/no assumption about the data
Currently, `curvlinops` assumes that `X` and `y` are both `torch.Tensor`. This works for the old deep learning paradigm. But with the rise of LLMs and other complicated models, one should make that assumption. 

For instance, the input of a Huggingface model is a `UserDict`:
```python
data = UserDict({
    'input_ids': torch.LongTensor(...),
    'attention_mask': torch.LongTensor(...),
    'labels': torch.LongTensor(...)
})
```
In this case, one can extract `X` and `y` via (example from [`laplace-torch`](https://github.com/aleximmer/Laplace/blob/2f1860bfeadbc933e91347a694be8ab7dda2e1f7/laplace/baselaplace.py#L465-L467))
```python
if isinstance(data, UserDict)  or isinstance(data, dict): # To support Huggingface dataset
    X, y = data, data['labels'].to(self._device)
```

However, `curvlinops` is strongly assuming `X` to be a tensor, see for example [here](https://github.com/f-dangel/curvlinops/blob/5852711aedf2728bc609fabfa95eac00da1beb63/curvlinops/_base.py#L330-L332) and [here](https://github.com/f-dangel/curvlinops/blob/5852711aedf2728bc609fabfa95eac00da1beb63/curvlinops/_base.py#L107-L111) and [here](https://github.com/f-dangel/curvlinops/blob/5852711aedf2728bc609fabfa95eac00da1beb63/curvlinops/_base.py#L371-L373).

I think the best way to circumvent this is to not assume anything about `X` (preferably also `y`, e.g. for multi-output models).

### PR DESCRIPTION
**Pull request by wiseodd**
Closes #86

@f-dangel @runame please give feedback and answer the question below. 

***Assumption:***

``` python
data: Union[Iterable[Tuple[Tensor, Tensor]], Iterable[Tuple[UserDict, Tensor]], Iterable[Tuple[dict, Tensor]]]
```
and there is an additional parameter in `_base._LinearOperator`:
``` python
batch_size_fn: Optional[Callable[[Any], int]] = None
```
where it must be non-`None` whenever `X` is not a `torch.Tensor`. 

This also fits well with Huggingface, although one must replace HF's default dataloader to outputs `(data, data['labels'])` instead of just `data`. Let me know if this should be considered.

***Code for testing the functionality:***

https://gist.github.com/wiseodd/426061afae24199446e60bfabc00e26e
I use `laplace-torch` there (two birds one stone), so just remove it if you don't want to install it. If you want to test `laplace-torch`, install via 
``` bash
pip install git+https://github.com/aleximmer/Laplace.git@mc-subset2
```

### CODE REVIEW

**Comment by wiseodd** [comment]
Just to link this to https://github.com/aleximmer/Laplace/pull/144

**Comment by wiseodd** [comment]
@f-dangel @runame I added (i) an example on the usage with HuggingFace transformers in the docs, and (ii) unit tests.

Please review.

**Comment by wiseodd** [comment]
Here's an example use case. 

https://curvlinops--100.org.readthedocs.build/en/100/basic_usage/example_huggingface.html

The load for the users is minimal and the choice of `UserDict` is compatible with HF. Note that the users can do whatever they want in terms of their `X`'s since the only requirements are they are `dict`/`UserDict` and they handle their "preprocessing" like `.to(device)` inside the `forward` function of the model.

File: docs/examples/basic_usage/example_huggingface.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
+r"""Usage with Huggingface LLMs
+===============================
+
+This example demonstrates how to work with Huggingface (HF) language models.
+
+As always, let's first import the required functionality.
+Remember to run `pip install -U transformers datasets`
```
Comment (f-dangel) [inline docs/examples/basic_usage/example_huggingface.py:7]: I think the correct rst syntax would be  
```
:code:`pip install ...`
```

File: docs/examples/basic_usage/example_huggingface.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+# make deterministic
+torch.manual_seed(0)
+numpy.random.seed(0)
+
+# %%
+#
+# Data
+# ----
+#
+# We will use synthetic data for simplicty. But obviously this can
```
Comment (f-dangel) [inline docs/examples/basic_usage/example_huggingface.py:36]: simplicity

File: docs/examples/basic_usage/example_huggingface.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+# be replace with any HF dataloader.
```
Comment (f-dangel) [inline docs/examples/basic_usage/example_huggingface.py:37]: replaced

File: docs/examples/basic_usage/example_huggingface.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    print(k, v.shape)
+
+
+# %%
+#
+# Model
+# -----
+#
+# Curvlinops supports general :code:`UserDict` inputs. However, everything must
+# be handled inside inside the :code:`forward` function of the model. This gives
```
Comment (f-dangel) [inline docs/examples/basic_usage/example_huggingface.py:78]: duplicate 'inside'

File: docs/examples/basic_usage/example_huggingface.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+        )
+
+        # For simplicity, only enable grad for the last layer
+        for p in self.hf_model.parameters():
+            p.requires_grad = False
+
+        for p in self.hf_model.score.parameters():
+            p.requires_grad = True
+
+    def forward(self, data: UserDict):
```
Comment (f-dangel) [inline docs/examples/basic_usage/example_huggingface.py:101]: Add docstrings

File: test/cases.py [hunk 6]
```diff
@@ -1,5 +1,6 @@
 """Contains test cases for linear operators."""
 
+from torch.functional import Tensor
```
Comment (f-dangel) [inline test/cases.py:3]: Maybe just `from torch import Tensor`?

File: test/cases.py [hunk 7]
```diff
@@ -208,3 +223,70 @@ def data():
...
+    {
+        "model_func": lambda: ModelWithDictInput(num_classes=2),
+        "loss_func": lambda: MSELoss(reduction="mean"),
+        "data": lambda: [
+            ({"x": rand(3, 10)}, regression_targets((3, 2))),
+            ({"x": rand(4, 10)}, regression_targets((4, 2))),
+        ],
+        "seed": 0,
+    },
+]
```
Comment (f-dangel) [inline test/cases.py:286]: Could just append to `CASES_NO_DEVICE`, then we can proceed without having to duplicate the test functions.

**Review by f-dangel — COMMENTED** [review]
Did a first pass and added some minor suggestions, mainly missing documentation.
One main suggestion is to avoid having to duplicate the test functions, simply by appending the test cases using a `dict` to `CASES_NO_DEVICE` and instead make the `cases` fixture return the `batch_size_fn`.

File: curvlinops/_base.py [hunk 8]
```diff
@@ -33,12 +34,13 @@ def __init__(
         model_func: Callable[[Tensor], Tensor],
         loss_func: Union[Callable[[Tensor, Tensor], Tensor], None],
         params: List[Parameter],
-        data: Iterable[Tuple[Tensor, Tensor]],
+        data: Union[Iterable[Tuple[Tensor, Tensor]], Iterable[Tuple[UserDict, Tensor]], Iterable[Tuple[dict, Tensor]]],
```
Comment (runame) [inline curvlinops/_base.py:37]: Shorter and maybe easier to read:
```python
Iterable[Tuple[Union[Tensor, UserDict, dict], Tensor]]
```

File: curvlinops/fisher.py [hunk 9]
```diff
@@ -109,12 +111,13 @@ def __init__(
         model_func: Callable[[Tensor], Tensor],
         loss_func: Union[MSELoss, CrossEntropyLoss],
         params: List[Parameter],
-        data: Iterable[Tuple[Tensor, Tensor]],
+        data: Union[Iterable[Tuple[Tensor, Tensor]], Iterable[Tuple[UserDict, Tensor]], Iterable[Tuple[dict, Tensor]]],
```
[the remark this answers is not in the document]
Comment (runame) [reply curvlinops/fisher.py:114]: Same comment as in `_base.py`.

File: curvlinops/fisher.py [hunk 10]
```diff
@@ -166,6 +171,7 @@ def __init__(
             progressbar=progressbar,
             check_deterministic=check_deterministic,
             num_data=num_data,
+            batch_size_fn=batch_size_fn
```
Comment (runame) [inline curvlinops/fisher.py:174]: Add trailing comma

File: curvlinops/kfac.py [hunk 11]
```diff
@@ -111,7 +113,7 @@ def __init__(
         model_func: Module,
         loss_func: MSELoss,
         params: List[Parameter],
-        data: Iterable[Tuple[Tensor, Tensor]],
+        data: Union[Iterable[Tuple[Tensor, Tensor]], Iterable[Tuple[UserDict, Tensor]], Iterable[Tuple[dict, Tensor]]],
```
[the remark this answers is not in the document]
Comment (runame) [reply curvlinops/kfac.py:116]: Same comment as `_base.py`.

File: curvlinops/kfac.py [hunk 12]
```diff
@@ -257,6 +262,7 @@ def __init__(
             check_deterministic=check_deterministic,
             shape=shape,
             num_data=num_data,
+            batch_size_fn=batch_size_fn
```
Comment (runame) [inline curvlinops/kfac.py:265]: Add trailing comma

File: test/cases.py [hunk 13]
```diff
@@ -20,9 +22,22 @@
...
 DEVICES_IDS = [f"dev={d}" for d in DEVICES]
 
+
+class ModelWithDictInput(Module):
+    def __init__(self, num_classes=2):
+        super().__init__()
+        self.net = Sequential(Linear(10, 5), ReLU(), Linear(5, num_classes))
+
+    def forward(self, data: UserDict | dict):
+        x = data["x"]
```
Comment (runame) [inline test/cases.py:37]: When `torch.cuda.is_available()` this model has to transfer to device here, otherwise the test will fail I think.

File: test/test_fisher.py [hunk 14]
```diff
@@ -78,3 +78,17 @@ def test_LinearOperator_matmat_expectation(
...
+    n_params = sum([p.numel() for p in params])
+
+    with raises(ValueError):
+        op = FisherMCLinearOperator(model_func, loss_func, params, data)
+
+    batch_size_fn = lambda data: data["x"].shape[0]
+    op = FisherMCLinearOperator(
+        model_func, loss_func, params, data, batch_size_fn=batch_size_fn
+    )
+    assert(op.shape == (n_params, n_params))
```
Comment (runame) [inline test/test_fisher.py:94]: I think we should apply the linop to a vector or matrix here as well (but as Felix said, if you add the test cases to the others, we don't have to duplicate the whole test anyway).

File: test/test_ggn.py [hunk 15]
```diff
@@ -33,3 +34,17 @@ def test_GGNLinearOperator_matmat(case, adjoint: bool, num_vecs: int = 3):
...
+    n_params = sum([p.numel() for p in params])
+
+    with raises(ValueError):
+        op = GGNLinearOperator(model_func, loss_func, params, data)
+
+    batch_size_fn = lambda data: data["x"].shape[0]
+    op = GGNLinearOperator(
+        model_func, loss_func, params, data, batch_size_fn=batch_size_fn
+    )
+    assert(op.shape == (n_params, n_params))
```
Comment (runame) [inline test/test_ggn.py:50]: Same comment as for `test_FisherLinearOperator_dict`.

File: test/test_gradient_moments.py [hunk 16]
```diff
@@ -25,3 +27,16 @@ def test_EFLinearOperator_matmat(case, adjoint: bool, num_vecs: int = 3):
...
+    n_params = sum([p.numel() for p in params])
+
+    with raises(ValueError):
+        op = EFLinearOperator(model_func, loss_func, params, data)
+
+    batch_size_fn = lambda data: data["x"].shape[0]
+    op = EFLinearOperator(
+        model_func, loss_func, params, data, batch_size_fn=batch_size_fn
+    )
+    assert(op.shape == (n_params, n_params))
```
Comment (runame) [inline test/test_gradient_moments.py:42]: Same comment as for `test_FisherLinearOperator_dict`.

File: test/test_hessian.py [hunk 17]
```diff
@@ -67,3 +67,17 @@ def test_blocked_HessianLinearOperator_matmat(
...
+    n_params = sum([p.numel() for p in params])
+
+    with raises(ValueError):
+        op = HessianLinearOperator(model_func, loss_func, params, data)
+
+    batch_size_fn = lambda data: data["x"].shape[0]
+    op = HessianLinearOperator(
+        model_func, loss_func, params, data, batch_size_fn=batch_size_fn
+    )
+    assert(op.shape == (n_params, n_params))
```
Comment (runame) [inline test/test_hessian.py:83]: Same comment as for `test_FisherLinearOperator_dict`.

File: test/test_kfac.py [hunk 18]
```diff
@@ -995,3 +995,17 @@ def test_forward_only_fisher_type_exact_weight_sharing_case(
...
+    n_params = sum([p.numel() for p in params])
+
+    with raises(ValueError):
+        op = KFACLinearOperator(model_func, loss_func, params, data)
+
+    batch_size_fn = lambda data: data["x"].shape[0]
+    op = KFACLinearOperator(
+        model_func, loss_func, params, data, batch_size_fn=batch_size_fn
+    )
+    assert(op.shape == (n_params, n_params))
```
Comment (runame) [inline test/test_kfac.py:1011]: Same comment as for `test_FisherLinearOperator_dict`.

**Review by runame — COMMENTED** [review]
While for many settings the model/data loader still has to be adjusted, this seems like a nice usability improvement!

See my comments and formatting and linting have to be fixed as well.

**Comment by runame** [comment]
One more thing, the type hints for the data also have to be modified in all files that don't require other changes, e.g. `ggn.py`.

File: test/cases.py (hunk 7 above)
Comment (wiseodd) [reply test/cases.py:286]: There are a couple of reasons:
1. We don't need to retest everything --- the only thing that needs to be tested is to validate that dict-like inputs are accepted. I made no other changes beyond the specification of the model's inputs.
2. One must change `examples/functorch.py` to reuse the old test cases.

@f-dangel @runame let me know what you think. I'm fine with either way but leaning towards separating the `DICT_CASES` since the purpose is just to validate the input format, and retest every other cases seem excessive.

Comment (f-dangel) [reply test/cases.py:286]: Re 1: I would propose to significantly reduce the number of test cases when using dict-based data loaders.
Re 2: Since we are adding a new flag `batch_size_fn` to the base class of all linear operators, it will make sense to introduce this argument into all functions related to testing/examples.

**Comment by wiseodd** [comment]
I'm finally done.

1. `examples/functorch.py` is now aware of dict-like inputs
2. Update *all* the test cases to accommodate `batch_size_fn`

@f-dangel @runame ready for another check.

File: docs/examples/basic_usage/example_huggingface.py [hunk 19]
```diff
@@ -0,0 +1,175 @@
...
+ggn = GGNLinearOperator(
+    model,
+    nn.CrossEntropyLoss(),
+    params,
+    [(data, data["labels"])],  # We still need to input a list of "(X, y)" pairs!
+    check_deterministic=False,
+    batch_size_fn=batch_size_fn,  # Remember to specify this!
+)
+
+G = ggn @ torch.eye(ggn.shape[0])
```
[the remark this answers is not in the document]
Comment (f-dangel) [reply docs/examples/basic_usage/example_huggingface.py:163]: (`curvlinops` multiplies onto `scipy/numpy` arrays)

File: test/conftest.py [hunk 20]
```diff
@@ -32,7 +34,15 @@ def initialize_case(
...
     data = case["data"]()
 
-    return model_func, loss_func, params, data
+    batch_size_fn = None
+
+    try:
+        if isinstance(next(iter(data))[0], MutableMapping):
+            batch_size_fn = lambda datum: datum["x"].shape[0]
+    except KeyError:  # The case of KFAC, where there are expand and reduce keys
+        batch_size_fn = None
```
Comment (f-dangel) [inline test/conftest.py:43]: It's subtle to see how the `try` block causes a `KeyError`.  Can you explain or rewrite the `if` statment?

File: test/test__base.py [hunk 21]
```diff
@@ -7,7 +7,9 @@
 
 def test_check_deterministic(non_deterministic_case):
     """Test that non-deterministic behavior is recognized."""
-    model_func, loss_func, params, data = non_deterministic_case
+    model_func, loss_func, params, data, batch_size_fn = non_deterministic_case
 
     with raises(RuntimeError):
-        _LinearOperator(model_func, loss_func, params, data, check_deterministic=True)
+        _LinearOperator(
+            model_func, loss_func, params, data, batch_size_fn, check_deterministic=True
```
Comment (f-dangel) [inline test/test__base.py:14]: use `batch_size_fn=batch_size_fn` to make sure this does not break if we introduce more keyword arguments.

File: test/test_ggn.py [hunk 22]
```diff
@@ -22,11 +32,13 @@ def test_GGNLinearOperator_matvec(case, adjoint: bool):
...
-    model_func, loss_func, params, data = case
+    model_func, loss_func, params, data, batch_size_fn = case
 
-    op = GGNLinearOperator(model_func, loss_func, params, data)
+    op = GGNLinearOperator(
+        model_func, loss_func, params, data, batch_size_fn=batch_size_fn
+    )
     op_functorch = (
-        functorch_ggn(model_func, loss_func, params, data).detach().cpu().numpy()
+        functorch_ggn(model_func, loss_func, params, data, "x").detach().cpu().numpy()
```
Comment (f-dangel) [inline test/test_ggn.py:41]: Specify `input_key='x'` to improve readability.

File: test/test_gradient_moments.py [hunk 23]
```diff
@@ -18,8 +37,19 @@ def test_EFLinearOperator_matvec(case, adjoint: bool):
...
-    op = EFLinearOperator(*case)
-    op_functorch = functorch_empirical_fisher(*case).detach().cpu().numpy()
+    model_func, loss_func, params, data, batch_size_fn = case
+
+    op = EFLinearOperator(
+        model_func, loss_func, params, data, batch_size_fn=batch_size_fn
+    )
+    op_functorch = (
+        functorch_empirical_fisher(
+            model_func, loss_func, params, data, batch_size_fn, "x"
```
Comment (f-dangel) [inline test/test_gradient_moments.py:47]: Specify `input_key='x'` and `batch_size_fn=batch_size_fn` to improve readability.

File: test/test_hessian.py [hunk 24]
```diff
@@ -11,8 +13,22 @@
...
+    # Test when X is dict-like but batch_size_fn = None (default)
+    if isinstance(data[0][0], MutableMapping):
+        with raises(ValueError):
+            op = HessianLinearOperator(model_func, loss_func, params, data)
+
+    op = HessianLinearOperator(
+        model_func, loss_func, params, data, batch_size_fn=batch_size_fn
+    )
+    op_functorch = (
+        functorch_hessian(model_func, loss_func, params, data, "x")
```
Comment (f-dangel) [inline test/test_hessian.py:27]: Specify `input_key='x'` to improve readability.

File: test/test_hessian.py [hunk 25]
```diff
@@ -21,8 +37,17 @@ def test_HessianLinearOperator_matvec(case, adjoint: bool):
...
 def test_HessianLinearOperator_matmat(case, adjoint: bool, num_vecs: int = 3):
-    op = HessianLinearOperator(*case)
-    op_functorch = functorch_hessian(*case).detach().cpu().numpy()
+    model_func, loss_func, params, data, batch_size_fn = case
+
+    op = HessianLinearOperator(
+        model_func, loss_func, params, data, batch_size_fn=batch_size_fn
+    )
+    op_functorch = (
+        functorch_hessian(model_func, loss_func, params, data, "x")
```
Comment (f-dangel) [inline test/test_hessian.py:46]: Specify `input_key='x'` to improve readability.

Comment (f-dangel) [inline test/test_hessian.py:46]: Specify `input_key='x'` to improve readability.

Comment (f-dangel) [inline test/test_hessian.py:46]: Specify `input_key='x'` to improve readability.

File: test/test_hessian.py [hunk 26]
```diff
@@ -50,14 +75,21 @@ def test_blocked_HessianLinearOperator_matmat(
...
+        params,
+        data,
+        batch_size_fn=batch_size_fn,
+        block_sizes=block_sizes,
+    )
 
     # compute the blocks with functorch and build the block diagonal matrix
     op_functorch = [
-        functorch_hessian(model, loss_func, params_block, data).detach()
+        functorch_hessian(model_func, loss_func, params_block, data, "x").detach()
```
Comment (f-dangel) [inline test/test_hessian.py:92]: Specify `input_key='x'` to improve readability.

File: test/test_inverse.py [hunk 27]
```diff
@@ -27,14 +28,16 @@
...
-    GGN = GGNLinearOperator(model_func, loss_func, params, data)
+    GGN = GGNLinearOperator(
+        model_func, loss_func, params, data, batch_size_fn=batch_size_fn
+    )
     damping = aslinearoperator(delta * sparse.eye(GGN.shape[0]))
 
     inv_GGN = CGInverseLinearOperator(GGN + damping)
     inv_GGN_functorch = inv(
-        functorch_ggn(model_func, loss_func, params, data).detach().cpu().numpy()
+        functorch_ggn(model_func, loss_func, params, data, "x").detach().cpu().numpy()
```
Comment (f-dangel) [inline test/test_inverse.py:40]: Specify `input_key='x'` to improve readability.

File: test/test_inverse.py [hunk 28]
```diff
@@ -44,14 +47,16 @@ def test_CG_inverse_damped_GGN_matvec(case, delta: float = 2e-2):
...
```
Comment (f-dangel) [inline test/test_inverse.py:59]: Specify `input_key='x'` to improve readability.

File: test/test_inverse.py [hunk 29]
```diff
@@ -61,13 +66,15 @@ def test_CG_inverse_damped_GGN_matmat(case, delta: float = 1e-2, num_vecs: int =
...
 
-    GGN = GGNLinearOperator(model_func, loss_func, params, data)
+    GGN = GGNLinearOperator(
+        model_func, loss_func, params, data, batch_size_fn=batch_size_fn
+    )
     damping = aslinearoperator(delta * sparse.eye(GGN.shape[0]))
 
     damped_GGN_functorch = functorch_ggn(
-        model_func, loss_func, params, data
+        model_func, loss_func, params, data, "x"
```
Comment (f-dangel) [inline test/test_inverse.py:77]: Specify `input_key='x'` to improve readability.

File: test/utils.py [hunk 30]
```diff
@@ -77,14 +79,15 @@ def ggn_block_diagonal(
         loss_func: The loss function.
         params: The parameters w.r.t. which the GGN block-diagonals will be computed.
         data: A data loader.
+        batch_size_fn: A function that returns the batch size of given a dict-like ``X``.
```
Comment (f-dangel) [inline test/utils.py:82]: remove 'of'

File: test/test_jacobian.py [hunk 31]
```diff
@@ -1,17 +1,27 @@
...
+    # Test when X is dict-like but batch_size_fn = None (default)
+    if isinstance(data[0][0], MutableMapping):
+        with raises(AttributeError):
+            op = JacobianLinearOperator(model_func, params, data)
 
-    op = JacobianLinearOperator(model_func, params, data)
-    op_functorch = functorch_jacobian(model_func, params, data).detach().cpu().numpy()
+    op = JacobianLinearOperator(model_func, params, data, batch_size_fn=batch_size_fn)
+    op_functorch = (
+        functorch_jacobian(model_func, params, data, "x").detach().cpu().numpy()
```
Comment (f-dangel) [inline test/test_jacobian.py:23]: Specify `input_key='x'` to improve readability.

File: test/test_jacobian.py [hunk 32]
```diff
@@ -20,10 +30,12 @@ def test_JacobianLinearOperator_matvec(case, adjoint: bool):
...
 
 def test_JacobianLinearOperator_matmat(case, adjoint: bool, num_vecs: int = 3):
-    model_func, _, params, data = case
+    model_func, _, params, data, batch_size_fn = case
 
-    op = JacobianLinearOperator(model_func, params, data)
-    op_functorch = functorch_jacobian(model_func, params, data).detach().cpu().numpy()
+    op = JacobianLinearOperator(model_func, params, data, batch_size_fn=batch_size_fn)
+    op_functorch = (
+        functorch_jacobian(model_func, params, data, "x").detach().cpu().numpy()
```
Comment (f-dangel) [inline test/test_jacobian.py:37]: Specify `input_key='x'` to improve readability.

File: test/test_jacobian.py [hunk 33]
```diff
@@ -32,10 +44,19 @@ def test_JacobianLinearOperator_matmat(case, adjoint: bool, num_vecs: int = 3):
...
+    # Test when X is dict-like but batch_size_fn = None (default)
+    if isinstance(data[0][0], MutableMapping):
+        with raises(AttributeError):
+            op = TransposedJacobianLinearOperator(model_func, params, data)
+
+    op = TransposedJacobianLinearOperator(
+        model_func, params, data, batch_size_fn=batch_size_fn
+    )
+    op_functorch = (
+        functorch_jacobian(model_func, params, data, "x").detach().cpu().numpy().T
```
Comment (f-dangel) [inline test/test_jacobian.py:58]: Specify `input_key='x'` to improve readability.

File: test/test_jacobian.py [hunk 34]
```diff
@@ -46,10 +67,14 @@ def test_TransposedJacobianLinearOperator_matvec(case, adjoint: bool):
...
-
-    op = TransposedJacobianLinearOperator(model_func, params, data)
-    op_functorch = functorch_jacobian(model_func, params, data).detach().cpu().numpy().T
+    model_func, _, params, data, batch_size_fn = case
+
+    op = TransposedJacobianLinearOperator(
+        model_func, params, data, batch_size_fn=batch_size_fn
+    )
+    op_functorch = (
+        functorch_jacobian(model_func, params, data, "x").detach().cpu().numpy().T
```
Comment (f-dangel) [inline test/test_jacobian.py:76]: Specify `input_key='x'` to improve readability.

File: test/test_submatrix_on_curvatures.py [hunk 35]
```diff
@@ -58,15 +58,22 @@ def every_third_idxs(dim: int):
...
     col_idxs = submatrix_case["col_idx_fn"](dim)
 
-    A = operator_case(*case)
+    A = operator_case(model_func, loss_func, params, data, batch_size_fn=batch_size_fn)
     A_sub = SubmatrixLinearOperator(A, row_idxs, col_idxs)
 
-    A_functorch = CURVATURE_IN_FUNCTORCH[operator_case](*case)
+    if operator_case == EFLinearOperator:
+        A_functorch = CURVATURE_IN_FUNCTORCH[operator_case](
+            model_func, loss_func, params, data, batch_size_fn, "x"
```
Comment (f-dangel) [inline test/test_submatrix_on_curvatures.py:71]: Specify `input_key='x'` to improve readability.

File: test/test_submatrix_on_curvatures.py [hunk 36]
```diff
@@ -58,15 +58,22 @@ def every_third_idxs(dim: int):
...
+        )
+    else:
+        A_functorch = CURVATURE_IN_FUNCTORCH[operator_case](
+            model_func, loss_func, params, data, "x"
```
Comment (f-dangel) [inline test/test_submatrix_on_curvatures.py:75]: Specify `input_key='x'` to improve readability.

**Comment by wiseodd** [comment]
@f-dangel I'm done. Currently running the test on a GPU-enabled env. It takes so long but you can continue reviewing. PEP8 and other linters' issues should also be resolved.

**Comment by wiseodd** [comment]
Alright, confirmed that all tests passed on GPU!

**Comment by f-dangel** [comment]
@runame I'm done with my second pass, could you take a quick second look and merge if everything looks good?

**Review by runame — APPROVED** [review]
LGTM!
